internal/auth/oidc/auth_method.go

```diff
@@ -0,0 +1,193 @@
+			Description:  opts.withDescription,
+			State:        string(InactiveState),
+			DiscoveryUrl: u,
+			ClientId:     clientId,
+			ClientSecret: string(clientSecret),
+			MaxAge:       int32(opts.withMaxAge),
+		},
+	}
+	if err := a.validate(op); err != nil {
+		return nil, err // intentionally not wrapped.
+	}
+	if a.ClientSecretHmac != "" {
+		return nil, errors.New(errors.InvalidParameter, op, "client secret hmac should be empty")
+	}
+	return a, nil
+}
+
+// validate the AuthMethod.  On success, it will return nil. Since setting up an
+// OIDC auth method requires a dance with the IdP, where you're need X before you
+// can configure Y, we allow things like the discovery URL, client ID, client
+// secret, etc to be empty until the AuthMethod moves into a PublicActive state.
+// That means validate can't completely ensure the data is valid and ultimately
+// we must rely on the database constraints/triggers to ensure the AuthMethod's
+// data integrity.
+func (a *AuthMethod) validate(caller errors.Op) error {
+	if a.ScopeId == "" {
+		return errors.New(errors.InvalidParameter, caller, "missing scope id")
+	}
+	if !validState(a.State) {
+		return errors.New(errors.InvalidParameter, caller, fmt.Sprintf("invalid state: %s", a.State))
+	}
+	if a.DiscoveryUrl != "" {
+		if _, err := url.Parse(a.DiscoveryUrl); err != nil {
+			return errors.New(errors.InvalidParameter, caller, "not a valid discovery URL", errors.WithWrap(err))
+		}
+	}
+	if a.MaxAge < -1 {
+		return errors.New(errors.InvalidParameter, caller, "max age cannot be less than -1")
+	}
+	return nil
+}
+
+// AllocAuthMethod makes an empty one in memory
+func AllocAuthMethod() AuthMethod {
+	return AuthMethod{
+		AuthMethod: &store.AuthMethod{},
+	}
+}
+
+// Clone an AuthMethod.
+func (a *AuthMethod) Clone() *AuthMethod {
+	cp := proto.Clone(a.AuthMethod)
+	return &AuthMethod{
+		AuthMethod: cp.(*store.AuthMethod),
+	}
+}
+
+// TableName returns the table name.
+func (a *AuthMethod) TableName() string {
+	if a.tableName != "" {
+		return a.tableName
+	}
+	return DefaultAuthMethodTableName
+}
+
+// SetTableName sets the table name.
+func (a *AuthMethod) SetTableName(n string) {
+	a.tableName = n
+}
+
+// oplog will create oplog metadata for the AuthMethod.
+func (a *AuthMethod) oplog(op oplog.OpType) oplog.Metadata {
+	metadata := oplog.Metadata{
+		"resource-public-id": []string{a.GetPublicId()},
+		"resource-type":      []string{"oidc auth method"},
+		"op-type":            []string{op.String()},
+		"scope-id":           []string{a.ScopeId},
+	}
+	return metadata
+}
+
+// encrypt the auth method before writing it to the db
+func (a *AuthMethod) encrypt(ctx context.Context, cipher wrapping.Wrapper) error {
+	const op = "oidc.(AuthMethod).encrypt"
+	if err := structwrapping.WrapStruct(ctx, cipher, a.AuthMethod, nil); err != nil {
+		return errors.Wrap(err, op, errors.WithCode(errors.Encrypt))
+	}
+	a.KeyId = cipher.KeyID()
+	if err := a.hmacClientSecret(cipher); err != nil {
```

Comment:
What will happen when the client secret is an empty string?

Reply:
An hmac of the empty string will be generated and callers can use it to determine if the secret has changed in the future.